api/src/main/java/marquez/db/JobVersionDao.java

```diff
@@ -14,205 +14,303 @@
+    jobVersionInputs.forEach(
+        jobVersionInput -> {
+          jobVersionDao.upsertInputDatasetFor(
+              jobVersionRow.getUuid(), jobVersionInput.getDatasetUuid());
+        });
 
-    // Only set current version if completed
-    if (runState == RunState.COMPLETED) {
-      jobDao.updateVersion(jobRow.getUuid(), transitionedAt, jobVersion.getUuid());
-    }
-    jobVersionDao.updateLatestRun(jobVersion.getUuid(), transitionedAt, runUuid);
-    for (ExtendedDatasetVersionRow datasetVersionRow : inputs) {
-      jobVersionDao.upsertDatasetIoMapping(
-          jobVersion.getUuid(), datasetVersionRow.getDatasetUuid(), IoType.INPUT, jobRow.getUuid());
-    }
+    // Link the output datasets to the job version.
+    jobVersionOutputs.forEach(
+        jobVersionOutput -> {
+          jobVersionDao.upsertOutputDatasetFor(
+              jobVersionRow.getUuid(), jobVersionOutput.getDatasetUuid());
+        });
 
-    for (ExtendedDatasetVersionRow datasetVersionRow : outputs) {
-      jobVersionDao.upsertDatasetIoMapping(
-          jobVersion.getUuid(),
-          datasetVersionRow.getDatasetUuid(),
-          IoType.OUTPUT,
-          jobRow.getUuid());
-    }
+    // Link the job version to the run.
+    createRunDao().updateJobVersion(runUuid, jobVersionRow.getUuid());
 
-    return new JobVersionBag(jobRow, inputs, outputs, jobVersion);
-  }
+    // Link the run to the job version; multiple run instances may be linked to a job version.
+    jobVersionDao.updateLatestRunFor(jobVersionRow.getUuid(), transitionedAt, runUuid);
 
-  default UUID buildJobVersion(
-      String namespaceName,
-      String jobName,
-      List<ExtendedDatasetVersionRow> inputs,
-      List<ExtendedDatasetVersionRow> outputs,
-      Map context,
-      String location) {
-    final byte[] bytes =
-        VERSION_JOINER
-            .join(
-                namespaceName,
-                jobName,
-                inputs.stream().flatMap(JobVersionDao::idToStream).collect(joining(VERSION_DELIM)),
-                outputs.stream().flatMap(JobVersionDao::idToStream).collect(joining(VERSION_DELIM)),
-                location,
-                KV_JOINER.join(context))
-            .getBytes(UTF_8);
-    return UUID.nameUUIDFromBytes(bytes);
+    // Link the job version to the job only if the run is marked done and has transitioned into one
+    // of the following states: COMPLETED, ABORTED, or FAILED.
+    if (runState.isDone()) {
+      jobDao.updateVersionFor(jobRow.getUuid(), transitionedAt, jobVersionRow.getUuid());
+    }
+
+    return new BagOfJobVersionInfo(jobRow, jobVersionRow, jobVersionInputs, jobVersionOutputs);
   }
 
-  static Stream<String> idToStream(ExtendedDatasetVersionRow dataset) {
-    return Stream.of(dataset.getNamespaceName(), dataset.getDatasetName());
+  /** Returns the specified {@link ExtendedDatasetVersionRow}s as {@link DatasetId}s. */
+  default ImmutableSet<DatasetId> toDatasetIds(
```

Comment:
Since order matters in the collection iteration, why not return a `List` or a `SortedSet`?

Reply:
Yeah, `SortedSet` makes sense here since `DatasetIds` are now comparable.